File: server/finances_explorer/models/member.py
import json

from flask_restplus import fields

from finances_explorer.restplus import api


class Member:
    get_member_response_model = api.model('Member', {
        'member_id': fields.Integer(example=1, description="The ID of the member in the database"),
        'title': fields.String(example="Mr.", description="Personal Title"),
        'name': fields.String(description="The member's full name"),
        'gender': fields.Integer(example=0, description="0-Masculino, 1-Feminino"),
        'nationality': fields.String(description="The member's nationality"),
        'birth_date': fields.Date(example="2019-01-01", description="Birth Date"),
        'death_date': fields.Date(example="2019-01-01", description="The death date of the member"),
        'street_name': fields.String(description="Member's address field: street"),
        'door_number': fields.String(description="Member's address field: door number"),
        'flat': fields.String(description="Member's address field: flat number"),
        'postal-code': fields.String(description="Member's address field: postal code"),
        'city': fields.String(description="Member's address field: city"),
        'contact1': fields.String(description="Member's main contact"),
        'contact2': fields.String(description="Member's secondary contact"),
        'email': fields.String(),
        'baptism_date': fields.Date(example="2019-01-01", description="The date of the member's baptism"),
        'communion_date': fields.Date(example="2019-01-01", description="The date of the member's communion"),
        'confirmation_date': fields.Date(example="2019-01-01", description="The date of the member's confirmation"),
        'marital_status': fields.String(),
        'partner_id': fields.Integer(example=1, description="The ID of the member's partner, when also a member"),
        'occupation': fields.String(description="The member's profession"),
        'professional_qualifications': fields.String(description="The highest academic degree level achieved"),
        'father_id': fields.Integer(example=2, description="The ID of the member's father, when also a member"),
        'mother_id': fields.Integer(example=3, description="The ID of the member's mother, when also a member")
    })

    post_member_request_model = api.model('Create new Member', {
        'title': fields.String(example="Mr.", description="Personal Title"),
        'name': fields.String(required=True, description="The member's full name"),
        'gender': fields.Integer(required=True, example=0, description="0-Masculino, 1-Feminino"),
        'nationality': fields.String(description="The member's nationality"),
        'birth_date': fields.Date(required=True, example="2019-01-01", description="Birth Date"),
        'death_date': fields.Date(example="2019-01-01", description="The death date of the member"),
        'street_name': fields.String(description="Member's address field: street"),
        'door_number': fields.String(description="Member's address field: door number"),
        'flat': fields.String(description="Member's address field: flat number"),
        'postal_code': fields.String(description="Member's address field: postal code"),
        'city': fields.String(description="Member's address field: city"),
        'contact1': fields.String(description="Member's main contact"),
        'contact2': fields.String(description="Member's secondary contact"),
        'email': fields.String(),
        'baptism_date': fields.Date(example="2019-01-01", description="The date of the member's baptism"),
        'communion_date': fields.Date(example="2019-01-01", description="The date of the member's communion"),
        'confirmation_date': fields.Date(example="2019-01-01", description="The date of the member's confirmation"),
        'marital_status': fields.String(),
        'partner_id': fields.Integer(example=1, description="The ID of the member's partner, when also a member"),
        'occupation': fields.String(description="The member's profession"),
        'professional_qualifications': fields.String(description="The highest academic degree level achieved"),
        'father_id': fields.Integer(example=2, description="The ID of the member's father, when also a member"),
        'mother_id': fields.Integer(example=3, description="The ID of the member's mother, when also a member")
    })

    def __init__(self, member_id="", title="", name="", gender="", nationality="", birth_date="",
                 death_date="", street_name="", door_number="", flat="", postal_code="", city="",
                 contact1="", contact2="", email="", baptism_date="", communion_date="", confirmation_date="",
                 marital_status="", partner_id="", occupation="", professional_qualifications="", father_id="",
                 mother_id=""):
        self.member_id = member_id
        self.title = title
        self.name = name
        self.gender = gender
        self.nationality = nationality
        self.birth_date = birth_date
        self.death_date = death_date
        self.street_name = street_name
        self.door_number = door_number
        self.flat = flat
        self.postal_code = postal_code
        self.city = city
        self.contact1 = contact1
        self.contact2 = contact2
        self.email = email
        self.baptism_date = baptism_date
        self.communion_date = communion_date
        self.confirmation_date = confirmation_date
        self.marital_status = marital_status
        self.partner_id = partner_id
        self.occupation = occupation
        self.professional_qualifications = professional_qualifications
        self.father_id = father_id
        self.mother_id = mother_id

    def __str__(self):
        attrs = vars(self)
        return ', '.join("%s: %s" % item for item in attrs.items())

    def to_json_string(self):
        return json.dumps(self.__dict__)

    def to_json(self):
        return json.loads(self.to_json_string())

File: app/src/app/modules/members/models/member.ts

export class Member {
    id: string;
    title: string;
    name: string;
    gender: string;
    nationality: string;
    birth_date: Date;
    current_age: Number;
    death_date: Date;
    street_name: string;
    door_number: string;
    flat: string;
    postal_code: string;
    city: string;
    contact1: string;
    contact2: string;
    email: string;
    baptism_date: Date;
    communion_date: Date;
    confirmation_date: Date;
    marital_status: string;
    partner_id: string;
    partner_name: string;
    occupation: string;
    professional_qualifications: string;
    father_id: string;
    father_name: string;
    mother_id: string;
    mother_name: string;

    public static map(data: any): Member[] {

        return data.map((member: any) => {

            var partner: Member = data.find(m => m.member_id == member.partner_id);
            var partner_name_aux = partner ? partner.name : null;

            var father: Member = data.find(m => m.member_id == member.father_id);
            var father_name_aux = father ? father.name : null;

            var mother: Member = data.find(m => m.member_id == member.mother_id);
            var mother_name_aux = mother ? mother.name : null;

            var timeDiff = Math.abs(Date.now() - (new Date(member.birth_date)).getTime());
            var age = Math.floor((timeDiff / (1000 * 3600 * 24))/365.25);

            var nationality = null;
            switch(member.nationality) {
                case "1":nationality = "Portuguesa"; break;
                case "2":nationality = "Brasileira"; break;
                case "3":nationality = "Americana"; break;
                case "4":nationality = "Espanhola"; break;
                case "5":nationality = "Inglesa"; break;
                case "6":nationality = "Francesa"; break;
                case "7":nationality = "Alemã"; break;
                case "8":nationality = "Japonesa"; break;
                case "9":nationality = "Chinesa"; break;
                case "10":nationality = "Angolana"; break;
                default:nationality = "Portuguesa"; break;
            }

            return {
                id: member.member_id,
                title: member.title,
                name: member.name,
                gender: member.gender,
                nationality: member.nationality,
                nationality_description: nationality,
                birth_date: new Date(member.birth_date),
                current_age: age,
                death_date: member.death_date,
                street_name: member.street_name,
                door_number: member.door_number,
                flat: member.flat,
                postal_code: member.postal_code,
                city: member.city,
                contact1: member.contact1,
                contact2: member.contact2,
                email: member.email,
                baptism_date: member.baptism_date,
                communion_date: member.communion_date,
                confirmation_date: member.confirmation_date,
                marital_status: member.marital_status,
                partner_id: member.partner_id,
                partner_name: partner_name_aux,
                occupation: member.occupation,
                professional_qualifications: member.professional_qualifications,
                father_id: member.father_id,
                father_name: father_name_aux,
                mother_id: member.mother_id,
                mother_name: mother_name_aux,
            };
        });
    }

}
